flashinfer/kda_evolution.py

```
@@ -0,0 +1,288 @@
+        if any(tensor.dtype != torch.bfloat16 for tensor in (q, k, v, g, beta)):
+            raise TypeError("q, k, v, g, and beta must use torch.bfloat16")
+        batch, tokens_per_batch, num_heads, _head_dim = q.shape
+        total_tokens = batch * tokens_per_batch
+        fixed_layout = cu_seqlens is None
+        if fixed_layout:
+            offsets = tuple(range(0, total_tokens + 1, tokens_per_batch))
+            cu_seqlens = torch.tensor(offsets, dtype=torch.int64, device=q.device)
+        else:
+            offsets = tuple(int(value) for value in cu_seqlens.tolist())
+        sequence_lengths = tuple(
+            end - start for start, end in zip(offsets, offsets[1:], strict=False)
```

Ruff (0.16.2)

[warning] Prefer `itertools.pairwise()` over `zip()` when iterating over successive pairs

Replace `zip()` with `itertools.pairwise()`

(RUF007)